### PR TITLE
 Ensure that txnId is the same if the request is retried

### PR DESCRIPTION
**Pull request by bmarty**
Complement to #3707 

Ensure that txnId is the same if the request is retried.

Also improve the value of generated requestId. We now use UUIDs

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/CancelGossipRequestWorker.kt [hunk 1]
```diff
@@ -43,6 +43,9 @@ internal class CancelGossipRequestWorker(context: Context,
             override val sessionId: String,
             val requestId: String,
             val recipients: Map<String, List<String>>,
+            // The txnId for the sendToDevice request. Nullable for compatibility reason, but MUST always be provided
```
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/CancelGossipRequestWorker.kt:46]:
```suggestion
            // The txnId for the sendToDevice request. Nullable for compatibility reasons, but MUST always be provided
```

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipRequestWorker.kt [hunk 2]
```diff
@@ -46,6 +46,9 @@ internal class SendGossipRequestWorker(context: Context,
             override val sessionId: String,
             val keyShareRequest: OutgoingRoomKeyRequest? = null,
             val secretShareRequest: OutgoingSecretRequest? = null,
+            // The txnId for the sendToDevice request. Nullable for compatibility reason, but MUST always be provided
```
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipRequestWorker.kt:49]:
```suggestion
            // The txnId for the sendToDevice request. Nullable for compatibility reasons, but MUST always be provided
```

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipRequestWorker.kt [hunk 3]
```diff
@@ -58,7 +61,8 @@ internal class SendGossipRequestWorker(context: Context,
     }
 
     override suspend fun doSafeWork(params: Params): Result {
-        val localId = LocalEcho.createLocalEchoId()
+        // (temporary code)
```
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipRequestWorker.kt:64]: Temporary how/why?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipWorker.kt [hunk 4]
```diff
@@ -48,6 +48,9 @@ internal class SendGossipWorker(context: Context,
             val requestUserId: String?,
             val requestDeviceId: String?,
             val requestId: String?,
+            // The txnId for the sendToDevice request. Nullable for compatibility reason, but MUST always be provided
```
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipWorker.kt:51]:
```suggestion
            // The txnId for the sendToDevice request. Nullable for compatibility reasons, but MUST always be provided
```

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipWorker.kt [hunk 5]
```diff
@@ -62,7 +65,8 @@ internal class SendGossipWorker(context: Context,
     }
 
     override suspend fun doSafeWork(params: Params): Result {
-        val localId = LocalEcho.createLocalEchoId()
+        // (temporary code)
```
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipWorker.kt:68]: Same as above, why is this temporary?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/tasks/SendToDeviceTask.kt [hunk 6]
```diff
@@ -46,16 +46,21 @@ internal class DefaultSendToDeviceTask @Inject constructor(
                 messages = params.contentMap.map
         )
 
+        // Create a unique txnId first, to use the same value if the request is retried
```
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/tasks/SendToDeviceTask.kt:49]: This comment is a bit hard to parse, also isn't it the other way around? We first try an existing transaction ID and if there isn't one we create a new random one?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/util/RequestIdHelper.kt [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.internal.crypto.util
+
+import java.util.UUID
+
+internal object RequestIdHelper {
+    fun createUniqueRequestId() = "req_" + UUID.randomUUID().toString()
```
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/util/RequestIdHelper.kt:22]: What's the point of prepending `req_` here? 

As far as I know everyone who uses a UUID for transaction IDs does so without modification. This will only allow servers to know which client is doing the request, avoiding the information leak is probably more important than whatever reason we have to do this.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/crypto/verification/VerificationService.kt [hunk 8]
```diff
@@ -52,7 +52,7 @@ interface VerificationService {
                              transactionId: String?): String?
 
     /**
-     * Request a key verification from another user using toDevice events.
+     * Request a key verification from another user using event in a room.
```
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/crypto/verification/VerificationService.kt:55]:
```suggestion
     * Request key verification with another user via room events (instead of the to-device API)
```

**Review by poljar — CHANGES_REQUESTED** [review]
I'm not quite as fluent in the EA codebase as I would like to give proper reviews of the functional change (i.e. does this code indeed make sure that the transaction id is reused when requests are retried), but the comments and the way the UUID is modified could do with some refinement.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/tasks/SendToDeviceTask.kt (hunk 6 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/tasks/SendToDeviceTask.kt:49]: I agree, I will update it

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipRequestWorker.kt (hunk 3 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipRequestWorker.kt:64]: params.txnId should be provided in all cases now. But `Params` can be deserialized by the WorkManager from data serialized in a previous version of the application, so without the `txnId`. That's why it has to stay nullable for a while... Probably until we delete all this code to switch to the RustSdk.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/util/RequestIdHelper.kt (hunk 7 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/util/RequestIdHelper.kt:22]: OK, removed

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipRequestWorker.kt (hunk 3 above)
Comment (poljar) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipRequestWorker.kt:64]: The comment is in this state isn't that helpful if that's going to hang around until the rust-sdk comes around. I would either expand it a bit to explain the state of affairs or just remove it to not confuse people that might be reading this code.

Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/SendGossipRequestWorker.kt]: You are right, I've updated the comment (at the 3 places).

**Review by poljar — APPROVED** [review]
Looks good from my point of view.
